git_repo/services/ext/gitbucket.py

```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env python
+
+import logging
+log = logging.getLogger('git_repo.gitbucket')
+
+from ..service import register_target
+from .github import GithubService
+
+import github3
+
+@register_target('bucket', 'gitbucket')
+class GitbucketService(GithubService):
+    fqdn = "localhost"
+    port = 8080
+    def __init__(self, *args, **kwarg):
+        super(GitbucketService, self).__init__(*args, **kwarg)
+
+    def connect(self):
+        self.gh = github3.GitHubEnterprise(self.build_url(self))
```

Comment:
cf [how I did it in gitlab](https://github.com/kounoike/git-repo/blob/e48a3b761102eb278351032a7c4b3fbc4c1034c2/git_repo/services/ext/gitlab.py#L24,L36).

The order is:

1. init phase:
  1. instanciate session (and stuff that go with it)
  2. load configuration
2. connect phase:
  1. if need be, overload the github reference (but preserve the session)
  2. do the connect stuff

basically here's the fixed gist:

```python
GITHUB_FQDN='github.com'

@register_target('hub', 'github')
class GithubService(RepositoryService):
    fqdn = GITHUB_FQDN

    def __init__(self, *args, **kwarg):
        self.gh = github3.GitHub()
        super(GithubService, self).__init__(*args, **kwarg)

    def connect(self):
        if self.fqdn != GITHUB_FQDN:
            # upgrade self.gh from a GitHub object to a GitHubEnterprise object
            gh = github3.GitHubEnterprise(RepositoryService.build_url(self))
            self.gh._session.base_url = gh._session.base_url
            self.gh = gh
            # propagate ssl certificate parameter
            self.gh._session.verify = self.session_certificate or not self.session_insecure
        try:
            self.gh.login(token=self._privatekey)
            self.username = self.gh.user().login
        except github3.models.GitHubError as err:
            if 401 == err.code:
                if not self._privatekey:
                    raise ConnectionError('Could not connect to Github. '
                                          'Please configure .gitconfig '
                                          'with your github private key.') from err
                else:
                    raise ConnectionError('Could not connect to Github. '
'Check your configuration and try again.') from err
```